aim/aim_store.py

```diff
@@ -329,60 +349,63 @@ def supports_hooks(self):
-        db_obj.setdefault('metadata', {'labels': {'aim_type': name}})
-        labels = db_obj['metadata']['labels']
-        attrs = db_obj['spec'][name]
-        for k, v in attribute_dict.iteritems():
-            if k in resource_klass.identity_attributes:
-                labels[k] = utils.sanitize_name(v)
-            attrs[k] = v
-        if 'name' not in db_obj['metadata']:
-            db_obj['metadata']['name'] = self._build_name(name, resource_klass,
-                                                          db_obj)
+        db_obj.from_attr(resource_klass, attribute_dict)
 
     def to_attr(self, resource_klass, db_obj):
-        result = {}
-        for k in resource_klass.attributes():
-            try:
-                result[k] = getattr(db_obj, k)
-            except AttributeError:
-                pass
-        return result
+        return db_obj.to_attr(resource_klass, defaults=self.attribute_defaults)
 
     def add(self, db_obj):
         created = None
-        try:
-            curr = self.klient.read_namespaced_aci(db_obj['metadata']['name'],
-                                                   self.namespace)
-            if curr:
-                # Replace this version object
-                curr['spec'] = db_obj['spec']
-                self.klient.replace_namespaced_aci(db_obj['metadata']['name'],
-                                                   self.namespace, curr)
-                created = curr
-        except api_v1.klient.ApiException as e:
-            if str(e.status) == '404':
-                # Object doesn't exist, create it.
-                db_obj.pop('resourceVersion', None)
-                self.klient.create_namespaced_aci(self.namespace, db_obj)
-                created = db_obj
-            else:
-                raise
+        k8s_klass = type(db_obj)
+        retries = 3  # this is arbitrary
+        while retries:
+            retries -= 1
+            try:
+                curr = self.klient.read(k8s_klass, db_obj['metadata']['name'],
+                                        self.namespace)
+                if curr:
+                    curr['spec'].update(db_obj.get('spec', {}))
+                    curr['metadata'].setdefault('annotations', {}).update(
+                        db_obj.get('metadata', {}).get('annotations', {}))
+                    curr['metadata'].setdefault('labels', {}).update(
+                        db_obj.get('metadata', {}).get('labels', {}))
+                    curr.pop('status', None)
+                    self.klient.replace(k8s_klass, db_obj['metadata']['name'],
+                                        self.namespace, curr)
+                    created = curr
+                break
+            except api_v1.klient.ApiException as e:
+                if str(e.status) == '404':
+                    # Object doesn't exist, create it.
+                    db_obj.pop('resourceVersion', None)
+                    self.klient.create(k8s_klass, self.namespace, db_obj)
```

Comment:
I don't think you would get a 409 (Conflict) for a create (HTTP POST), it is usually reported replace (HTTP PUT)